### PR TITLE
Support Connection API for testing possible connections

### LINKED ISSUE #179: Allow testing a new connection with a "dry run" type of request
Implement an endpoint that `vscode` can use before creating a new connection that would test the connection is valid instead of saving it. It would be used in the New Connection form UI as a "Test Connection" button (or similar), which is common in other connection form UIs

### PR DESCRIPTION
**Pull request by rhauch**
Resolves #179

## Summary of Changes

Expands the `POST /gateway/v1/connections` endpoint to add a `dry_run` query parameter that defaults to `false` for current behavior. When `dry_run=true`, the endpoint will not create a new connection but will instead return the Connection resource type response with a status that contains the information _as if_ the connection were created. This will allow the extension to have a “Test” button on the new connection panel, without having to actually create, wait for updated status, and then delete the connection.

The behavior of the existing endpoint will be changed as follows:
* When the `dry_run` query parameter is `false` (the default), the endpoint calls `ConnectionStateManager.createConnectionState(spec)` that checks for duplicate IDs (if provided), validates the input spec (failing if there are validation errors), adds the new connection in the `ConnectionStateManager`, and returns a Connection object with the supplied `ConnectionSpec` and _initial_ `ConnectionStatus`.
* When the `dry_run` query parameter is `true`, the endpoint calls a new method `ConnectionStateManager.testConnection(spec)` that checks for duplicate IDs (if provided), validates the input spec (failing if there are validation errors), and *synchronously* computes the status (via `ConnectionState.getConnectionStatus()`), and finally returns a Connection object with the supplied `ConnectionSpec` and computed `ConnectionStatus`.

## Any additional details or context that should be provided?

Found and fixed a bug in the `DirectConnectionState` that was incorrectly setting the `status.schema_registry.state` rather than null when there is no schema registry configured.

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

- Tests:
    - [x] Added new
    - [ ] Updated existing
    - [ ] Deleted existing
- [x] Have you validated this change locally against a running instance of the Quarkus dev server?
    ```shell
    make quarkus-dev
    ```
- [x] Have you validated this change against a locally running native executable?
    ```shell
    make mvn-package-native && ./target/ide-sidecar-*-runner
    ```

### CODE REVIEW

File: src/main/java/io/confluent/idesidecar/restapi/resources/ConnectionsResource.java [hunk 1]
```diff
@@ -83,11 +84,35 @@ public Uni<ConnectionsList> listConnections() {
...
+      // Get the status of the connection
+      var futureStatus = testedState.getConnectionStatus();
+      // And create a uni that will complete when the status is available
+      return Uni
+          .createFrom()
+          .completionStage(futureStatus.toCompletionStage())
+          .map(connectionStatus ->
+              Connection.from(testedState, connectionStatus)
+          );
+    }
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/resources/ConnectionsResource.java:108]: This is the bulk of the changes in the Connections API to add the `dry_run` query parameter to the `POST /gateway/v1/connections` endpoint.

File: src/main/java/io/confluent/idesidecar/restapi/resources/ConnectionsResource.java [hunk 2]
```diff
@@ -83,11 +84,35 @@ public Uni<ConnectionsList> listConnections() {
...
+    // Create a real connection state and return the connection using the initial status
+    var newState = connectionStateManager.createConnectionState(connectionSpec);
+    return Uni
+        .createFrom()
+        .item(
+            Connection.from(newState, newState.getInitialStatus())
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/resources/ConnectionsResource.java:114]: For real create-connection behavior, I changed this line to `Connection.from(newState)` to this, and removed the `Connection.from(ConnectionState state)` (without a status) to help ensure we always pass in the correct state.

File: src/main/java/io/confluent/idesidecar/restapi/connections/DirectConnectionState.java [hunk 3]
```diff
@@ -221,13 +218,8 @@ protected Future<SchemaRegistryStatus> getSchemaRegistryConnectionStatus() {
...
     }).orElseGet(() -> {
-      // There is no Schema Registry configuration, so return the NONE state
-      return Future.succeededFuture(
-          ConnectionStatusSchemaRegistryStatusBuilder
-              .builder()
-              .state(ConnectedState.NONE)
-              .build()
-          );
+      // There is no Schema Registry configuration, so return no status for SR
+      return Future.succeededFuture(null);
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/connections/DirectConnectionState.java:222]: This is the bug I found and fixed: when there is no `spec.schema_registry`, then we want `status.schema_registry` to also be null.

File: src/main/java/io/confluent/idesidecar/restapi/connections/DirectConnectionState.java [hunk 4]
```diff
@@ -177,13 +179,8 @@ protected Future<KafkaClusterStatus> getKafkaConnectionStatus() {
...
         () -> {
-            // There is no Kafka cluster configuration, so return the NONE state
-            return Future.succeededFuture(
-                ConnectionStatusKafkaClusterStatusBuilder
-                    .builder()
-                    .state(ConnectedState.NONE)
-                    .build()
-            );
+            // There is no Kafka cluster configuration, so return a null Kafka status
+            return Future.succeededFuture(null);
```
Comment (rhauch) [inline src/main/java/io/confluent/idesidecar/restapi/connections/DirectConnectionState.java:183]: And here, too: when there is no `spec.kafka_cluster`, then we want `status.kafka_cluster` to also be null.

File: src/test/java/io/confluent/idesidecar/restapi/integration/connection/DirectConnectionSuite.java [hunk 5]
```diff
@@ -14,12 +15,98 @@
...
+        .body("api_version", equalTo("gateway/v1"))
+        .body("kind", equalTo("Connection"))
+        .body("metadata.self", startsWith("http://localhost:26637/gateway/v1/connections/"))
+        .body("metadata.resource_name", nullValue())
+        .body("id", equalTo(spec.id()))
+        .body("spec.id", equalTo(spec.id()))
+        .body("spec.name", equalTo(spec.name()))
+        .body("spec.type", equalTo(ConnectionType.DIRECT.name()))
+        .body("status.authentication.status", equalTo(Status.NO_TOKEN.name()))
+        .body("status.kafka_cluster.state", equalTo(ConnectedState.SUCCESS.name()));
```
Comment (rhauch) [inline src/test/java/io/confluent/idesidecar/restapi/integration/connection/DirectConnectionSuite.java:45]: I've come to really like this style of verification of the response objects. IMO, it's very clear as to what the test is asserting.

File: src/main/java/io/confluent/idesidecar/restapi/resources/ConnectionsResource.java (hunk 1 above)
Comment (flippingbits) [reply src/main/java/io/confluent/idesidecar/restapi/resources/ConnectionsResource.java:108]: Kudos for making this look straightforward 👍

**Review by flippingbits — APPROVED** [review]
LGTM, @rhauch! Feel free to decide if you want to incorporate the minor suggestions.

**Comment by rhauch** [comment]
Fixed the build failures by merging IT fixes from `main`. New build is enqueued, and should start soon.

**Review by flippingbits — DISMISSED** [review]
Thanks for the update, @rhauch. LGTM! 💯

**Comment by rhauch** [comment]
It's about time the build passes. :-(

**Comment by rohitsanj** [comment]
🤦

**Review by rohitsanj — APPROVED** [review]
Trusting @flippingbits and re-approving. (Take 2)

**Pull request merged**